Acquire-python API reference (#1)
acquire-project/acquire-docs: docs/PythonAPI.md

# API Reference

> Information on the classes in Acquire-python along with the attributes and methods associated with them.

Reviewer: ```suggestion
Information on the classes in `acquire-imaging` along with the attributes and methods associated with them.
```
We really should think about renaming the repo.

## Class `AvailableData`

The `AvailableData` class represents the collection of frames that have been captured since the last call to runtime.get_available_data(). `AvailableData` objects should be set to have a short lifetime, since these objects reserve space on the video queue and will eventually block camera acquisition to ensure no data is overwritten before it can be processed.

```python
class AvailableData:
    def frames(self) -> Iterator[VideoFrame]:
        """Returns an iterator over the video frames in the available data."""
    
    def get_frame_count(self) -> int:
        """Returns the total number of video frames in the available data."""
    
    def __iter__(self) -> Iterator[VideoFrame]:
        """Returns an iterator over the video frames in the available data."""
```
- Call `get_frame_count()` to query the number of frames in an `AvailableData` object. 

> - The `frames` method provides an iterator over these frames.

Reviewer: Maybe add a note here that the `AvailableData` object itself can be iterated over because of the `__iter__` method.

- `settings`: An instance of the `CameraProperties` class which contains the settings for the camera.

- The `dict` method creates a dictionary of a `Camera` object's attributes.

## Class `CameraProperties`
The `CameraProperties` class is used to set the desired camera properties for acquisition.

```python
class CameraProperties:
    exposure_time_us: float
    line_interval_us: float
    binning: float
    pixel_type: SampleType
    readout_direction: Direction
    offset: Tuple[int, int]
    shape: Tuple[int, int]
    input_triggers: InputTriggers
    output_triggers: OutputTriggers

    def __init__(self, *args: None, **kwargs: Any) -> None: ...
    """Initializes a CameraProperties object with optional arguments."""
    
    def dict(self) -> Dict[str, Any]: ...
    """Returns a dictionary of the CameraProperties attributes."""
```

- `exposure_time_us`: How long in microseconds your camera should collect light from the sample. However, for simulated cameras, this is just a waiting period before generating the next frame.

> - `line_interval_us`: The time to scan one line in microseconds in a rolling shutter camera camera.

Reviewer: ```suggestion
- `line_interval_us`: The time to scan one line in microseconds in a rolling shutter camera.
```

- `binning`: How many adjacent pixels in each direction to combine by averaging. If binning is set to 1, no pixels will be combined.

> - `pixel_type`: An instance of the `SampleType` class which specifies the numerical data type.

Reviewer: Can you provide some examples here?

- `readout_direction`: An instance of the `Direction` class which specifies whether the data is readout forwards or backwards.

> - `offset`: A tuple of two integers representing the dark counts of the camera.

Reviewer: This is the first time I've seen the term "dark count". Is that standard? Based on what Google tells me, I don't think this is correct.

Reviewer 2: I think this should be the (x, y) offset of the image region of interest on the sensor, rather the dark count. For spinnaker and genicam, the values represent pixels after binning.

> - `shape`: A tuple of two integers representing the shape of the camera detector in pixels.

Reviewer: What do the values in the tuple represent?

Reviewer 2: Like offset, I think this describes the shape of the image region of interest. Again, the values represent pixels after binning for spinnaker/genicam.

- `input_triggers`: An instance of the `InputTriggers` class.

> - `output_triggers`: An instance of the `OutputTriggers` class.

Reviewer: Can this be more detailed?

```python
class DeviceIdentifier:
    id: Tuple[int, int]
    kind: DeviceKind
    name: str

    def __init__(self, *args: None, **kwargs: Any) -> None: ...
    """Initializes a DeviceIdentifier object with optional arguments."""

    def dict(self) -> Dict[str, Any]: ...
    """Returns a dictionary of the DeviceIdentifier attributes."""

    @staticmethod
    def none() -> DeviceIdentifier: ...
    """Returns a "None" type DeviceIdentifier. Useful when a DeviceIdentifier is not needed."""

    def __eq__(self, other: object) -> bool:
        """Checks if two DeviceIdentifier objects are equal."""
        
    def __ge__(self, other: object) -> bool:
        """Checks if this DeviceIdentifier is greater than or equal to another."""
        
    def __gt__(self, other: object) -> bool:
        """Checks if this DeviceIdentifier is greater than another."""
        
    def __le__(self, other: object) -> bool:
        """Checks if this DeviceIdentifier is less than or equal to another."""
        
    def __lt__(self, other: object) -> bool:
        """Checks if this DeviceIdentifier is less than another."""
        
    def __ne__(self, other: object) -> bool:

>         """Checks if two DeviceIdentifier objects are not equal."""

Reviewer: I don't know if these are necessary. It's probably enough to say somewhere that `DeviceIdentifier`s can be compared and ordered.

Author: Noted. I'm going to leave this as-is for now in case we move forward with incorporating these as doc string in the pyi file.

```

> - `id`: A tuple containing two integers that serve as the device ID.

Reviewer: What do the integers represent?

Author: I am not sure. All I could find is that they are u8 integers. Here is where I was looking: https://github.com/acquire-project/acquire-python/blob/cb2dd9bdec0f483c3391b54ef8996c570614379d/src/device.rs#L92

```python
class DeviceKind:
    Camera: ClassVar[DeviceKind] = DeviceKind.Camera
    NONE: ClassVar[DeviceKind] = DeviceKind.NONE
    Signals: ClassVar[DeviceKind] = DeviceKind.Signals
    StageAxis: ClassVar[DeviceKind] = DeviceKind.StageAxis
    Storage: ClassVar[DeviceKind] = DeviceKind.Storage

    def __init__(self, *args: None, **kwargs: Any) -> None:
        """Initializes the DeviceKind class."""

    def __eq__(self, other: object) -> bool:
        """Checks if two DeviceKind objects are equal."""
        
    def __ge__(self, other: object) -> bool:
        """Checks if this DeviceKind is greater than or equal to another."""
        
    def __gt__(self, other: object) -> bool:
        """Checks if this DeviceKind is greater than another."""
        
    def __int__(self) -> int:
        """Converts the DeviceKind to an integer."""
        
    def __le__(self, other: object) -> bool:
        """Checks if this DeviceKind is less than or equal to another."""
        
    def __lt__(self, other: object) -> bool:
        """Checks if this DeviceKind is less than another."""
        
    def __ne__(self, other: object) -> bool:

>         """Checks if two DeviceKind objects are not equal."""

Reviewer: Same thing here, it's probably enough to say that the `DeviceKind` can be ordered and compared.

Author: Noted. I'm going to leave this as-is for now in case we move forward with incorporating these as doc string in the pyi file.

```

- `Camera`: Class variable of `DeviceKind` that defines the cameras supported by the system. 

- `NONE`: Class variable of `DeviceKind` that is set to None if no device of the specified kind is available.

- `Signals`: Class variable of `DeviceKind` that defines the signals supported by the system.

- `StageAxis`: Class variable of `DeviceKind` that defines the stage axes supported by the system.

> - `Storage`: Class variable of `DeviceKind` that defines the storage supported by the system.

Reviewer: While technically class variables, these are actually enum values.

Author: Noted. I changed the language to "Enum-type class variable"

Author: Class variables are defined in this file as:
```
Attribute: ClassVariable[Class] = Class.Attribute
```
Are all those class variables in the api of enum-type?

## Class `DeviceManager` 

The `DeviceManager` class manages selection of available devices in the system.

```python
class DeviceManager:
    def devices(self) -> List[DeviceIdentifier]:
        """Returns a list of all available device identifiers."""
    
    @overload
    def select(self, kind: DeviceKind) -> Optional[DeviceIdentifier]:

>         """Selects a specified device.

Reviewer: ```suggestion
        """Selects the first available device of `kind`.
```

        
        Args:
            kind (DeviceKind): The type of device to select.
            
        Returns:

>             Optional[DeviceIdentifier]: The selected device identifier, or None if no device of the specified kind is available.

Reviewer: ```suggestion
            Optional[DeviceIdentifier]: The identifier of the first available device of `kind`, or `None` if no such device is available.
```

        """
    
    @overload
    def select(self, kind: DeviceKind, name: Optional[str]) -> Optional[DeviceIdentifier]:
        """Selects a specified device.
        
        Args:
            kind (DeviceKind): The type of device to select.
            name (Optional[str]): The name of the device to select.
            
        Returns:
            Optional[DeviceIdentifier]: The selected device identifier, or None if the specified device is not available.
        """
    
    def select_one_of(self, kind: DeviceKind, names: List[str]) -> Optional[DeviceIdentifier]:

>         """Selects one of the devices of the specified kind from the given list of names.

Reviewer: How does it make this selection? Or put another way, what should I as the user expect when I use this method?

Author: Looking at: https://github.com/acquire-project/acquire-python/blob/cb2dd9bdec0f483c3391b54ef8996c570614379d/src/device_manager.rs#L69

Not sure if I am interpreting this correctly, but it seems to select the first device that fits within the specifications, so whichever one happened to be listed first.

            
        Returns:
            Properties: The updated configuration properties.
        """
    
    def start(self) -> None:
        """Starts the runtime, allowing it to collect data."""
    
    def stop(self) -> None:
        """Stops the runtime, ending data collection after the max number of frames is collected."""
    
    def abort(self) -> None:
        """Aborts the runtime, terminating it immediately."""
```

- Call `device_manager()` to return the `DeviceManager` object associated with this `Runtime` instance. 

- Call `get_available_data` with a specific `stream_id` to return the `AvailableData` associated with the `stream_id`.

- Call `get_configuration()` to return the `Properties` object associated with this `Runtime` instance.

- Call `get_state()` to return the `DeviceState` object associated with this `Runtime` instance.

- Call `set_configuration` with a `Properties` object to change the properties of this `Runtime` instance.

- Call `start()` to begin data acquisition.

- Call `stop()` to end data acquisition once the max number of frames specified in `config.video[0].max_frame_count` is collected.

- Call `abort()` to  `Runtime` instance.

> Note: If you set the maximum frame count to 0, `runtime.stop()` will hang forever. In order to terminate acquisition in this case, you need to call `runtime.abort()` instead, which will stop acquisition immediately. You can also use `runtime.abort()` if you set the maximum frame count to some nonzero value but you want to terminate acquisition before you reach that maximum frame count.

Reviewer: This is my mistake. While I believe this was true at one point, it's no longer true.

- `exposure_time_us`: How long in microseconds your camera should collect light from the sample. However, for simulated cameras, this is just a waiting period before generating the next frame.

- `line_interval_us`: The time to scan one line in microseconds in a rolling shutter camera.

> - `binning`: How many adjacent pixels in each direction to combine by averaging. For example, if `binning` is set to 2, a 2x2 sqaure of pixels will be combined by averaging. If `binning` is set to 1, no pixels will be combined.

Reviewer: ```suggestion
- `binning`: How many adjacent pixels in each direction to combine by averaging. For example, if `binning` is set to 2, a 2x2 square of pixels will be combined by averaging. If `binning` is set to 1, no pixels will be combined.
```
